src/main/java/crawlercommons/sitemaps/SiteMapParserSAX.java

```diff
@@ -211,13 +211,30 @@ public AbstractSiteMap parseSiteMap(String contentType, byte[] content, URL url)
             if (XML_MEDIA_TYPES.contains(mediaType)) {
                 return processXml(url, content);
             } else if (TEXT_MEDIA_TYPES.contains(mediaType)) {
-                return processText(url.toString(), content);
+                return processText(url, content);
             } else if (GZ_MEDIA_TYPES.contains(mediaType)) {
-                return processGzip(url, content);
-            } else {
-                mediaType = MEDIA_TYPE_REGISTRY.getSupertype(mediaType); // Check parent
-                return parseSiteMap(mediaType.toString(), content, url);
+                InputStream decompressed;
+                MediaType embeddedType;
+                try {
+                    decompressed = new GZIPInputStream(new ByteArrayInputStream(content));
+                    embeddedType = MediaType.parse(TIKA.detect(decompressed));
+                } catch (Exception e) {
+                    UnknownFormatException err = new UnknownFormatException("Failed to detect embedded MediaType of gzipped sitemap: " + url + ", caused by " + e);
+                    err.initCause(e);
+                    throw err;
+                }
+                if (XML_MEDIA_TYPES.contains(embeddedType)) {
```

Comment:
The repetition of code is not very elegant. What about having a separate recursive method for detecting the type instead and then an old fashioned switch statement to call the corresponding processX method? Any reason not to do that?